---
title: "RNASeq Project"
author: "J. Deseo"
date: "4/8/2024"
---

```{r setup}
### Edit this line so your notebook renders properly
knitr::opts_knit$set(root.dir = normalizePath("~/KGI Spring 2024/GENE 5130 - Bioinformatics in R/Project")) 

```

Load the libraries
```{r}
### Import the important libraries
library(dplyr)
library(readr)
library(DESeq2)
library(ggplot2)
library(ComplexHeatmap)
library(clusterProfiler)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(ggrepel)
library(EnhancedVolcano)
library(WGCNA)

```

Run the entire dseq program to find significantly expressed genes
```{r}
### Run this chunk to import the counts table and metadata into the environment
counts <- read.table('../data/my_project_counts2.txt', row.names = 1, header = TRUE)
metadata <- read.csv('../data/project_metadata.csv', row.names = 1, header = TRUE)

```


```{r counts table}
# Solution: https://www.biostars.org/p/366025/
# Exclude extra columns
counts <- counts[, -c(1:5)]

```


```{r}
## Change column names to match metadata
new_column_names <- c("SRR24206827", "SRR24206825", "SRR24206826", "SRR24206824")
colnames(counts) <- new_column_names

```

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, colData = metadata, design = ~ Treatment)
counts(dds)

### Subset your ddseq object
subset <-  dds[(rowSums(counts(dds)) >= 10),]

```

```{r}
filter_counts <- dds[rowSums(counts(dds))>10]
filter_counts

```

```{r}
### Fix the base to a preferred treatment
relevel(dds$Treatment, ref='PBS')

```

```{r}
### Run deseq2
deseq_ob <- DESeq(subset)

```

```{r}
countdds <- DESeq(filter_counts)
counts(countdds, normalized=T)

```

```{r}
#### Save the results to a new object
res <- results(deseq_ob, alpha = 0.05)
res

```

```{r}
plotMA(countdds)

```

```{r}
sigs.df <-  as.data.frame(res)
sigs.df$symbol <- mapIds(org.Mm.eg.db, keys= rownames(sigs.df), keytype = 'ENSEMBL', colum = "SYMBOL")

```

```{r}
EnhancedVolcano(sigs.df, x='log2FoldChange', y = 'padj', lab = sigs.df$symbol, ylim = c(0, -log10(10e-3)),pCutoff = 0.5)

```
```{r}
EnhancedVolcano(sigs.df, x='log2FoldChange', y='padj', lab = sigs.df$symbol)

```

## Heatmaps!

```{R}
diff.df <- as.data.frame(sigs.df)
diff.df <- diff.df %>%
  filter(padj < 0.05)

mat <- counts(deseq_ob, normalized = T)[rownames(diff.df),]
mat.z <- t(apply(mat,1, scale))
colnames(mat.z) <- colnames(mat)
Heatmap(mat.z, cluster_rows= T, cluster_columns= T, name = "Z-score", row_labels = diff.df[rownames(mat.z),]$symbol)

```

## Use dplyr to subset the table of significant values until you get a nice looking figure. 
(You cannot plot all genes at once so you have to subset in order to gain some insights from 
your data)

```{R}
diff.df <- as.data.frame(sigs.df)
diff.df <- diff.df %>%
  filter(padj < 0.05, log2FoldChange > 4, baseMean > 100)

mat <- counts(deseq_ob, normalized = T)[rownames(diff.df),]
mat.z <- t(apply(mat,1, scale))
colnames(mat.z) <- colnames(mat)
Heatmap(mat.z, cluster_rows= T, cluster_columns= T, name = "Z-score", row_labels = diff.df[rownames(mat.z),]$symbol)

```

## Gene Ontology Analysis

```{R}
genes_to_test <- rownames(diff.df)
GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Mm.eg.db", keyType = 'ENSEMBL', ont = "BP")
fit <- plot(barplot(GO_results, showCategory=10))

```
```{R}
library(gprofiler2) 
go_upreg <- gost(genes_to_test, organism = 'mmusculus', significant = T)
gostplot(go_upreg, interactive = T)

```
